ai/recommender/surprise_recommenders.py

```diff
@@ -38,31 +43,118 @@ class SurpriseRecommender(BaseRecommender):
         self.reader = Reader()
-        # algo is the online Surprise-based rec system
-        # if algo is None:
-        #     self.algo = KNNBasic() 
-        # else:
-        #     self.algo = algo
+
+        self.ml_type = ml_type
+
+        if metric is None:
+            logger.debug('metric is None, setting...')
+            self.metric='bal_accuracy' if self.ml_type=='classifier' else 'mse'
+        else:
+            self.metric = metric
+        assert(self.metric is not None)
+
+        logger.debug('self.algo_name: '+self.algo_name)
+        logger.debug('ml_type: '+self.ml_type)
+        logger.debug('metric: '+self.metric)
+
+        if filename is None:
+            filename = (
+                    'ai/recommender/saved/'
+                    + self.algo_name
+                    + '_' + self.ml_type 
+                    + '_' + self.metric 
+                    + '_pmlb_20200505'
+                    +'.pkl.gz')
+
+        super().__init__(ml_type, metric, ml_p, filename=filename,
+                knowledgebase=knowledgebase, random_state=random_state)
+
+        # note: anything set after super().__init() will change the field 
+        # of any recommender from a file
+
         
-        self.first_fit = True
         self.min_epochs = 10
         self.max_epochs = 100
     
     @property
     def algo_name(self):
+        if type(self.algo).__name__ is None:
+            return type(self).__name__
         return type(self.algo).__name__
 
+    def set_training_data(self, results_data, results_mf=None, source='pennai'):
+        """Used for loading pickled recomenders to set results_df 
+        without training.
+        :param results_data: DataFrame with columns corresponding to:
+                'dataset'
+                'algorithm'
+                'parameters'
+                self.metric
+        :param results_mf: metafeatures for the datasets in results_data 
+        """
+        # update trained dataset models and hash table
+        super().update(results_data, results_mf, source)
+        self.update_training_data(results_data, shuffle=True)
+        rowHashes = hash_pandas_object(self.results_df).values 
+        newHash = hashlib.sha256(rowHashes).hexdigest()
+        if hasattr(self, 'results_df_hash'):
+            if newHash == self.results_df_hash:
+                logger.info('results_df hashes match')
+            else:
+                logger.warn('the results_df hash from the pickle is different')
+
+    def load(self, filename=None, knowledgebase = None):
+        """Load a saved recommender state."""
+        if knowledgebase is None:
+            logger.warning('A knowledgebase needs to be provided to load '
+            'Surprise Recommenders from file. Not loading.')
+            return 
+        loaded = super().load(filename=filename)
+        if loaded:
+            logger.info('setting training data...')
+            self.set_training_data(knowledgebase, source='knowledgebase')
```

Comment:
knowledgebase was defined as string in ai/recommender/base.py but `set_training_data` function below need `result_df` a pandas.DataFrame instead of a string.

Reply:
It seems "results_df" is needed in `self.set_training_data`

Reply:
self.results_df is set in `self.update_training_data()` which is called by `set_training_data`. "results_df_hash" is stored in the pickle during the call to `save`.

Reply:
Check line 101 in the `surprise_recommenders.py`. `resuld_data` is argument without assignemnt.

Reply:
yes, result_data needs to be provided to use set_training_data. i'm not entirely sure what you're asking, but I added documentation and renamed `set_training_data` to `_reconstruct_training_data` to be more clear.